File: 2016/22/solution.go
package main

import (
	"container/heap"
	"fmt"
	"math"
	"os"
	"path/filepath"
	"regexp"
	"runtime"
	"strconv"
	"strings"

	"github.com/weitecklee/adventofcode/utils"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	input := strings.Split(string(data), "\n")
	nodes := parseInput(input)
	fmt.Println(part1(nodes))
	fmt.Println(part2(nodes))
}

type Node struct {
	pos   [2]int
	size  int
	used  int
	avail int
}

func (n *Node) Viability(n2 *Node) bool {
	return n.used > 0 && n.used <= n2.avail
}

func parseInput(input []string) []Node {
	nodes := []Node{}
	re := regexp.MustCompile(`\d+`)
	for i := 2; i < len(input); i++ {
		numsStr := re.FindAllString(input[i], -1)
		nums := []int{}
		for _, s := range numsStr {
			n, _ := strconv.Atoi(s)
			nums = append(nums, n)
		}
		nodes = append(nodes, Node{
			pos:   [2]int{nums[0], nums[1]},
			size:  nums[2],
			used:  nums[3],
			avail: nums[4],
		})
	}
	return nodes
}

func part1(nodes []Node) int {
	pairs := 0
	for i := 0; i < len(nodes); i++ {
		for j := i + 1; j < len(nodes); j++ {
			if nodes[i].Viability(&nodes[j]) {
				pairs++
			}
			if nodes[j].Viability(&nodes[i]) {
				pairs++
			}
		}
	}
	return pairs
}

type Value struct {
	steps int
	pos   [2]int
}

func dist(a, b [2]int) int {
	return int(math.Abs(float64(a[0]-b[0]))) + int(math.Abs(float64(a[1]-b[1])))
}

func part2(nodes []Node) int {

	nodeMap := map[[2]int]*Node{}
	var xMax, yMax int
	var emptyNode *Node
	for i := 0; i < len(nodes); i++ {
		nodeMap[nodes[i].pos] = &nodes[i]
		if nodes[i].pos[0] > xMax {
			xMax = nodes[i].pos[0]
		}
		if nodes[i].pos[1] > yMax {
			yMax = nodes[i].pos[1]
		}
		if nodes[i].used == 0 {
			emptyNode = &nodes[i]
		}
	}

	target := [2]int{xMax - 1, 0}
	steps := 0

	directions := [][2]int{
		{0, 1},
		{1, 0},
		{0, -1},
		{-1, 0},
	}

	queue := utils.NewMinHeap[Value]()

	heap.Push(queue, &utils.Item[Value]{
		Priority: dist(emptyNode.pos, target),
		Value:    Value{steps: 0, pos: emptyNode.pos},
	})
	visited := map[[2]int]int{}

	for len(queue.PriorityQueue) > 0 {
		item := heap.Pop(queue).(*utils.Item[Value])

		if item.Value.pos == target {
			steps = item.Value.steps
			break
		}

		curr, x, y := item.Value.steps, item.Value.pos[0], item.Value.pos[1]
		curr++

		for _, d := range directions {
			dx, dy := d[0], d[1]
			x2, y2 := x+dx, y+dy
			if x2 < 0 || x2 > xMax || y2 < 0 || y2 > yMax {
				continue
			}
			if nd, ok := nodeMap[[2]int{x2, y2}]; ok && nd.used > 100 {
				continue
			}
			if v, ok := visited[[2]int{x2, y2}]; ok && v <= curr {
				continue
			}
			visited[[2]int{x2, y2}] = curr
			heap.Push(queue, &utils.Item[Value]{Priority: dist([2]int{x2, y2}, target) + curr, Value: Value{steps: curr, pos: [2]int{x2, y2}}})
		}
	}

	return steps + 1 + (xMax-1)*5
}

File: 2016/22/solution.ts
import * as fs from "fs";
import * as path from "path";
import MinHeap from "../../utils/MinHeap";

const puzzleInput = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8")
  .split("\n");

const nodeRegex = /^(\S+)\s+(\d+)\w\s+(\d+)\w\s+(\d+)\w\s+(\d+)%$/;

class Node {
  name: string;
  size: number;
  used: number;
  avail: number;
  pos: number[];
  coords: string;
  constructor(line: string) {
    const parts = line.match(nodeRegex)!;
    this.name = parts[1];
    this.size = Number(parts[2]);
    this.used = Number(parts[3]);
    this.avail = Number(parts[4]);
    const coords = this.name.match(/\d+/g);
    this.pos = coords!.map(Number);
    this.coords = this.pos.join(",");
  }
}

const nodes = puzzleInput.slice(2).map((a) => new Node(a));

function part1(): number {
  let res = 0;
  for (let i = 0; i < nodes.length; i++) {
    for (let j = i + 1; j < nodes.length; j++) {
      if (nodes[i].used > 0 && nodes[i].used <= nodes[j].avail) res++;
      if (nodes[j].used > 0 && nodes[j].used <= nodes[i].avail) res++;
    }
  }
  return res;
}

// The solution method is hinted at by the problem description.
// All nodes (except for one empty node) have more data than there is space
// available on any other node so you cannot combine data from multiple nodes
// in one node. The only way is to keep swapping the empty node with another.
// There are some outlier nodes, which are very large and very full. Those
// are effectively walls and must be maneuvered around.
// As shown in the problem description, locate the empty node, move it to
// the left of the target node, then move the target node by swapping and
// shifting the empty node around it, all while avoiding the wall nodes.
// Calculate shortest route from empty node to left of target node, swap with
// target node, then start the following process:
// - repeatedly move to left of [current node holding target data] by moving
//   around it and then swap, as illustrated in problem description
// (This assumes there are no wall nodes that interfere in first two rows)
// Each cycle of this process takes 5 steps.
// Resulting number of steps is sum of:
// - shortest route from empty node to left of target node
// - 1 for first swap
// - 5 * (maximum x coordinate - 1)   (because it just got swapped)

function part2(): number {
  let steps = 0;
  const nodeMap: Map<string, Node> = nodes.reduce((a, b) => {
    a.set(b.coords, b);
    return a;
  }, new Map());
  const xMax = Math.max(...nodes.map((n) => n.pos[0]));
  const yMax = Math.max(...nodes.map((n) => n.pos[1]));

  // confirm there are no wall nodes in top two rows(y = 0 and y = 1)
  for (let x = 0; x <= xMax; x++) {
    if (nodeMap.get(`${x},0`)!.used > 100) return -1;
    if (nodeMap.get(`${x},1`)!.used > 100) return -1;
  }

  let emptyPos: number[] = [];
  for (const node of nodes) {
    if (node.used === 0) {
      emptyPos = node.pos;
      break;
    }
  }

  const directions = [
    [-1, 0],
    [1, 0],
    [0, 1],
    [0, -1],
  ];

  function heuristic(x: number, y: number): number {
    return Math.abs(xMax - 1 - x) + y;
  }

  const queue: [number, ...number[]][] = [[0, 0, emptyPos[0], emptyPos[1]]];
  const visited: Map<string, number> = new Map();
  while (queue.length) {
    let [_, curr, x, y] = MinHeap.pop(queue) as number[];
    if (x === xMax - 1 && y === 0) {
      steps = curr;
      break;
    }
    curr++;
    for (const [dx, dy] of directions) {
      const [x2, y2] = [x + dx, y + dy];
      if (x2 < 0 || y2 < 0 || x2 > xMax || y2 > yMax) continue;
      const node = nodeMap.get(`${x2},${y2}`)!;
      if (node.used > 100) continue;
      if (visited.has(node.coords) && visited.get(node.coords)! <= curr)
        continue;
      visited.set(node.coords, curr);
      MinHeap.push(queue, [curr + heuristic(x2, y2), curr, x2, y2]);
    }
  }

  return steps + 1 + (xMax - 1) * 5;
}

console.log(part1());
console.log(part2());
